internal/oci/notation/notation.go

```diff
@@ -0,0 +1,319 @@
+func WithLogger(logger logr.Logger) Options {
+	return func(o *options) {
+		o.Logger = logger
+	}
+}
+
+// NotationVerifier is a struct which is responsible for executing verification logic
+type NotationVerifier struct {
+	auth     authn.Authenticator
+	keychain authn.Keychain
+	verifier *notation.Verifier
+	opts     []remote.Option
+	insecure bool
+	logger   logr.Logger
+}
+
+type trustStore struct {
+	cert []byte
+}
+
+// GetCertificates implements truststore.X509TrustStore.
+func (s trustStore) GetCertificates(ctx context.Context, storeType truststore.Type, namedStore string) ([]*x509.Certificate, error) {
+	raw := s.cert
+	block, _ := pem.Decode(raw)
+	if block != nil {
+		raw = block.Bytes
+	}
+
+	certs, err := x509.ParseCertificates(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse certificate '%s': %s", namedStore, err)
+	}
+
+	return certs, nil
+}
+
+// NewNotationVerifier initializes a new Verifier
+func NewNotationVerifier(opts ...Options) (*NotationVerifier, error) {
+	o := options{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	store := &trustStore{
+		cert: o.RootCertificate,
+	}
+
+	verifier, err := verifier.New(o.TrustPolicy, store, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &NotationVerifier{
+		auth:     o.Auth,
+		keychain: o.Keychain,
+		verifier: &verifier,
+		opts:     o.ROpt,
+		insecure: o.Insecure,
+		logger:   o.Logger,
+	}, nil
+}
+
+// Verify verifies the authenticity of the given ref OCI image.
+// It returns a boolean indicating if the verification was successful.
+// It returns an error if the verification fails, nil otherwise.
+func (v *NotationVerifier) Verify(ctx context.Context, ref name.Reference) (oci.VerificationResult, error) {
+	url := ref.Name()
+
+	remoteRepo, err := v.remoteRepo(url)
+	if err != nil {
+		return oci.VerificationResultFailed, err
+	}
+
+	repo := registry.NewRepository(remoteRepo)
+
+	repoUrl, err := v.repoUrlWithDigest(url, ref)
+	if err != nil {
+		return oci.VerificationResultFailed, err
+	}
+
+	verifyOptions := notation.VerifyOptions{
+		ArtifactReference:    repoUrl,
+		MaxSignatureAttempts: 3,
+	}
+
+	_, outcomes, err := notation.Verify(ctx, *v.verifier, repo, verifyOptions)
+	if err != nil {
+		return oci.VerificationResultFailed, err
+	}
+
+	if len(outcomes) == 0 {
+		return oci.VerificationResultFailed, fmt.Errorf("signature verification failed for all the signatures associated with %s", url)
+	}
+
+	outcome := outcomes[0]
+
+	if outcome.VerificationLevel == trustpolicy.LevelSkip {
+		return oci.VerificationResultIgnored, nil
+	}
+
+	for _, i := range outcome.VerificationResults {
+		if i.Error != nil {
+			if i.Type == trustpolicy.TypeAuthenticity {
+				return oci.VerificationResultIgnored, i.Error
+			}
+
+			if i.Action == trustpolicy.ActionLog {
+				v.logger.Info(fmt.Sprintf("verification check for type %s failed for %s with message %s", i.Type, url, i.Error.Error()))
+			}
+		}
+	}
+
+	return oci.VerificationResultSuccess, nil
+}
```

Comment:
This is great I think. Thanks.

The logic is quite specific though, we need a test for this to guard against future changes. Also maybe a comment to explain the logic.

Reply:
Comments add in [0dedf6f](https://github.com/fluxcd/source-controller/pull/1075/commits/0dedf6f098a9897a3bc2e54306d77af6bf15b449) and moved to separate function + tests in [c3a89f4](https://github.com/fluxcd/source-controller/pull/1075/commits/c3a89f4025768101f0d9ae813bb1916a53959cb3).